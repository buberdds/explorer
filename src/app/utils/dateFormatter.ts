import {
  secondsInDay,
  secondsInHour,
  secondsInMinute,
  secondsInMonth,
  secondsInWeek,
  secondsInYear,
} from 'date-fns/constants'
import { differenceInSeconds } from 'date-fns/differenceInSeconds'
import { intlFormatDistance } from 'date-fns/intlFormatDistance'

const dateFormat = new Intl.DateTimeFormat()

export const intlDateFormat = (date: Date | number) => dateFormat.format(date)

// TODO: Works only in en-US locale, as suffixes are hardcoded
export const formatDistanceToNow = (
  date: Date | number,
  baseDate: Date | number = new Date(),
  locale = 'en-US',
) => {
  const diffInSeconds = differenceInSeconds(date, baseDate)
  let unit: Intl.RelativeTimeFormatUnit
  // Simplified from date-fns code, but without quarters
  if (Math.abs(diffInSeconds) < secondsInMinute) {
    unit = 'second'
  } else if (Math.abs(diffInSeconds) < secondsInHour) {
    unit = 'minute'
  } else if (Math.abs(diffInSeconds) < secondsInDay) {
    unit = 'hour'
  } else if (Math.abs(diffInSeconds) < secondsInWeek) {
    unit = 'day'
  } else if (Math.abs(diffInSeconds) < secondsInMonth) {
    unit = 'week'
  } else if (Math.abs(diffInSeconds) < secondsInYear) {
    unit = 'month'
  } else {
    unit = 'year'
  }

  return intlFormatDistance(date, baseDate, {
    unit,
    style: 'short',
    numeric: 'always',
    locale,
  })
    .replace(/ ago$/, '')
    .replace(/^in /, '')
    .replace(/\.$/, '')
}
