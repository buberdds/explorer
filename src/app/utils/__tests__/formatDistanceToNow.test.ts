import { formatDistanceToNow } from '../dateFormatter'

describe('formatDistanceToNow', () => {
  beforeEach(() => {
    jest.useFakeTimers()
    jest.setSystemTime(new Date('2023-01-01T01:01:01Z'))
  })

  afterEach(() => {
    jest.useRealTimers()
  })

  it('should format future dates', () => {
    expect([
      formatDistanceToNow(new Date('2023-01-01T01:01:01Z')),
      formatDistanceToNow(new Date('2023-01-01T01:01:21Z')),
      formatDistanceToNow(new Date('2023-01-01T01:21:21Z')),
      formatDistanceToNow(new Date('2023-01-01T02:01:01Z')),
      formatDistanceToNow(new Date('2023-01-02T01:01:01Z')),
      formatDistanceToNow(new Date('2023-02-01T01:01:01Z')),
      formatDistanceToNow(new Date('2024-01-01T01:01:01Z')),
      formatDistanceToNow(new Date('2023-08-23T12:20:56Z')),
    ]).toMatchSnapshot()
  })

  it('should format past dates', () => {
    expect([
      formatDistanceToNow(new Date('2023-01-01T01:01:01Z')),
      formatDistanceToNow(new Date('2023-01-01T01:00:41Z')),
      formatDistanceToNow(new Date('2023-01-01T00:41:01Z')),
      formatDistanceToNow(new Date('2023-01-01T00:01:01Z')),
      formatDistanceToNow(new Date('2022-12-31T01:01:01Z')),
      formatDistanceToNow(new Date('2022-12-01T01:01:01Z')),
      formatDistanceToNow(new Date('2022-01-01T01:01:01Z')),
      formatDistanceToNow(new Date('2022-04-23T12:20:56Z')),
    ]).toMatchSnapshot()
  })
})
