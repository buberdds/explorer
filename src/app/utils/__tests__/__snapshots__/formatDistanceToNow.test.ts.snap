// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`formatDistanceToNow should format future dates 1`] = `
[
  "0 sec",
  "20 sec",
  "20 min",
  "1 hr",
  "1 day",
  "1 mo",
  "12 mo",
  "7 mo",
]
`;

exports[`formatDistanceToNow should format past dates 1`] = `
[
  "0 sec",
  "20 sec",
  "20 min",
  "1 hr",
  "1 day",
  "1 mo",
  "12 mo",
  "9 mo",
]
`;
